app/scripts/snaps/preinstalled-snaps.ts

```diff
@@ -0,0 +1,57 @@
+import type { PreinstalledSnap } from '@metamask/snaps-controllers';
+
+// Preinstalled snaps requires fs and require to read and use any installed snaps.
+// We can switch `require` to `import`, but then this file gets transpiled to ESM & won't have access to `require.resolve`.
+// Work is needed to create the require object in ESM - specifically the import url. See snippet
+// import { createRequire } from 'node:module'
+// const require = createRequire(import.meta.url)
+//                                ^ The 'import.meta' meta-property is not allowed in files which will build into CommonJS output
+/* eslint-disable-next-line @typescript-eslint/no-require-imports,@typescript-eslint/no-var-requires */
+const fs = require('fs');
+
+const TEMP_PREINSTALLED_SNAPS = [];
+
+///: BEGIN:ONLY_INCLUDE_IF(snaps)
+try {
+  TEMP_PREINSTALLED_SNAPS.push(
+    getPreinstalledSnap(
+      '@metamask/message-signing-snap',
+      fs.readFileSync(
+        require.resolve('@metamask/message-signing-snap/snap.manifest.json'),
+        'utf-8',
+      ),
+      [
+        {
+          path: 'images/icon.svg',
+          value: fs.readFileSync(
+            require.resolve('@metamask/message-signing-snap/images/icon.svg'),
+          ),
+        },
+        {
+          path: 'dist/bundle.js',
+          value: fs.readFileSync(
+            require.resolve('@metamask/message-signing-snap/dist/bundle.js'),
+          ),
+        },
+      ],
+    ),
+  );
+} catch (e) {
+  console.error('Failed to add pre-installed snap', e);
+}
+
+function getPreinstalledSnap(
+  npmPackage: string,
+  manifest: PreinstalledSnap['manifest'],
+  files: PreinstalledSnap['files'],
+): PreinstalledSnap {
+  return {
+    snapId: `npm:${npmPackage}` as PreinstalledSnap['snapId'],
+    manifest: JSON.parse(manifest),
+    files,
+    removable: false,
+  };
+}
+///: END:ONLY_INCLUDE_IF
+
+export const PREINSTALLED_SNAPS = Object.freeze(TEMP_PREINSTALLED_SNAPS);
```

Comment:
Question here: if we require `PREINSTALLED_SNAPS`, it is possible to be `[]` first and then the file processing happens, and then we have filled `TEMP_PREINSTALLED_SNAPS` with push. So correct me if i'm wrong, this racing condition could be an issue. Can we try this patter: export a function as `getPreinstalledSnaps` and use async/await to return the value instead?

Reply:
Unlike ESM `import()` syntax, `require` is synchronous, so is the `fs.readFileSync`.

Let treat this as a nit - we can refactor to make into an async function once we add more pre-installed snaps.

Reply:
@DDDDDanica did you find @Prithpal-Sooriya 's answer to this question satisfactory?

Reply:
@Prithpal-Sooriya can we have a ticket to track this refactor todo in order to not lose track of it? It is also helpful to add this todo ticket number in the codebase as well !

and @brad-decker yes, this is acceptable ~

Reply:
Preinstalled snaps are currently not asynchronous. The Snap source files need to be available when the `MetaMaskController` boots synchronously. The code that @Prithpal-Sooriya is proposing will mostly run at build-time to inline these required files in the build so that they can be used at runtime synchrnously.

I don't believe there is a race condition here.